### PR TITLE
Enhance triggers documentation

### PR DESCRIPTION
**Pull request by rbino**

### CODE REVIEW

File: doc/pages/architecture/060-triggers.md [hunk 1]
```diff
@@ -1,23 +1,374 @@
...
+- `value_change`: verifies the condition whenever a new sample is not equal to the last one.
+- `value_change_applied`: verifies the condition whenever a new sample is not equal to the last one,
+  after it is saved to the database.
+- `path_created`: verifies the condition whenever a new property interface is set or the first value
+  is streamed on a datastream interface.
+- `path_created`: verifies the condition whenever a property is unset.
+
+`interface_name` and `interface_major` represent, respectively, the Interface name and major version
+that uniquely identify an Astarte Interface. `interface_name` can be `*` to match all interfaces; in
+that case `interface_major` is ignored.
```
Comment (drf) [inline doc/pages/architecture/060-triggers.md]: Specify what happens here. All major numbers are matched?

File: doc/pages/architecture/060-triggers.md [hunk 2]
```diff
@@ -1,23 +1,374 @@
...
+
+## Actions
+
+Actions are triggered by a matching condition. An Action defines how the event should be sent to the
+outer world (e.g. an http POST on a certain URL). In addition, most actions have a Payload, which
+carries the body of the event.
+
+Payloads are most of the time represented as text, and Astarte provides several ways to generate
+them. By default Astarte generates a JSON payload with all the relevant information of the event.
+This is also the format used when delivering payloads in Astarte Channels. The format of the payload
```
Comment (drf) [inline doc/pages/architecture/060-triggers.md]: Cross-link Astarte Channels documentation here.

File: doc/pages/architecture/060-triggers.md [hunk 3]
```diff
@@ -1,23 +1,374 @@
...
+The payload of the request is JSON document with this format:
+```json
+{
+  "timestamp": "<timestamp>",
+  "device_id": "<device_id>",
+  "event": <event>
+}
+```
+
+`timestamp` is an ISO8601 timestamp representing when the event happened.
```
Comment (drf) [inline doc/pages/architecture/060-triggers.md]: Put here an example of a ISO8601 timestamp

File: doc/pages/architecture/060-triggers.md [hunk 4]
```diff
@@ -1,23 +1,374 @@
...
+
+The Mustache action sends an HTTP `POST` request to the specified `http_post_url`, with the payload
+built with the [Mustache](https://mustache.github.io/) template specified in `template`. If the
+template contains a key inside a double curly bracket (like so: `{{ key }}`), it will be substituted
+with the actual value of that key in the event.
+
+The basic keys that can be use to populate the template are:
+- `{{ realm }}`: the realm the trigger belongs to.
+- `{{ device_id }}`: the device that originated the trigger.
+- `{{ event_type}}`: the type of the received event.
```
Comment (drf) [inline doc/pages/architecture/060-triggers.md]: Cosmetics: missing a space before }}

File: doc/pages/architecture/060-triggers.md [hunk 5]
```diff
@@ -1,23 +1,374 @@
...
+
+#### Example
+
+This is an example of a trigger that uses Mustache action.
+```json
+{
+  "name": "example_mustache_trigger",
+  "action": {
+    "template_type": "mustache",
+    "template": "Device { device_id } just connected from IP { device_ip_address}",
```
Comment (drf) [inline doc/pages/architecture/060-triggers.md]: Cosmetics: missing space before }

File: doc/pages/user/060-using_triggers.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+Triggers allow receiving notifications when a device connects, disconnects or publishes specific
+data.
+
+This page will show two practical examples regarding triggers. For more detailed documentation,
+showing all possible trigger conditions and actions, check the [Triggers](060-triggers.html) page.
+
+## Connection Trigger
+
+This trigger will send a `POST` request to `<post-url>` every time any device connects to the MQTT
+broker.
```
Comment (drf) [inline doc/pages/user/060-using_triggers.md]: s/the MQTT broker/its Transport/

File: doc/pages/architecture/060-triggers.md (hunk 2 above)
Comment (rbino) [reply doc/pages/architecture/060-triggers.md]: The one in the User Guide? Currently there's no Channels documentation in the Architecture part besides the one at the bottom of this page

File: doc/pages/architecture/060-triggers.md [hunk 7]
```diff
@@ -1,23 +1,376 @@
...
+}
+```
 
-Payloads are most of the time plain-text, and Astarte provides several ways to generate them. By default Astarte generates a JSON payload with all the relevant information of the event. This is also the format used when delivering payloads in Astarte Channels. The format for each payload can be found in the [simple events encoder](https://github.com/astarte-platform/astarte_core/blob/master/lib/astarte_core/triggers/simple_events/encoder.ex). In the foreseeable future, more user friendly documentation about `json` payloads will be provided.
+The `condition` is represented by the `simple_triggers` array. In this release, Astarte supports
+only a single entry in the `simple_triggers` array, but support for multiple simple triggers (and
+different ways to combine them) is planned for future releases.
 
-Astarte also provides a powerful templating mechanism for plain-text payloads based on top of [Mustache](https://mustache.github.io/). This is especially useful for integrating with third-party actors which require custom payload formats.
+This specific `condition` specifies that when data is received on the `org.astarteplatform.Values`
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: This example `condition`

File: doc/pages/architecture/060-triggers.md [hunk 8]
```diff
@@ -1,23 +1,376 @@
...
+
+Data triggers are installed for all devices in a Realm.
+
+#### Parameters
+
+`data_trigger_type` can be one of the following:
+- `incoming_data`: verifies the condition whenever new data arrives.
+- `value_stored`: verifies the condition whenever new data arrives, after it is saved to the
+  database.
+- `value_change`: verifies the condition whenever the last sample received is different from the
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: the last data or last value. value_change is also valid for properties, so we don't use the word sample.

File: doc/pages/architecture/060-triggers.md [hunk 9]
```diff
@@ -1,23 +1,376 @@
...
+  last one previously received.
+- `value_change_applied`: verifies the condition whenever the last sample received is different from
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: last sample also here.

File: doc/pages/architecture/060-triggers.md [hunk 10]
```diff
@@ -1,23 +1,376 @@
...
+  the last one previously received, after it is saved to the database.
+- `path_created`: verifies the condition whenever a new path in a property interface is set or the
+  first value is streamed on a datastream interface.
+- `path_created`: verifies the condition whenever a property path is unset.
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: it looks like a copy & paste error

File: doc/pages/architecture/060-triggers.md [hunk 11]
```diff
@@ -1,23 +1,376 @@
...
+`interface_name` and `interface_major` represent, respectively, the Interface name and major version
+that uniquely identify an Astarte Interface. `interface_name` can be `*` to match all interfaces; in
+that case `interface_major` is ignored.
+
+`match_path` is the path that will be used to match the condition. It can be `/*` to match all the
+paths of an interface.
+
+`value_match_operator` is the operator used to match the incoming data against a known value. It can
+be `*` to indicate that all values should be matched (`known_value` is ignored in that case),
+otherwise it can be one of these operators: `==`, `!=`, `>`, `>=`, `<`, `<=`, `contains`,
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: contains applies only to strings,  binaryblobs and arrays. right?

File: doc/pages/architecture/060-triggers.md [hunk 12]
```diff
@@ -1,23 +1,376 @@
...
+`not_contains`. The match is always performed with `<incoming_value> <operator> <known_value>`.
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: not contains applies only to strings,  binaryblobs and arrays. right?

File: doc/pages/architecture/060-triggers.md [hunk 13]
```diff
@@ -1,23 +1,376 @@
...
+
+Payloads are most of the time represented as text, and Astarte provides several ways to generate
+them. By default Astarte generates a JSON payload with all the relevant information of the event.
+This is also the format used when delivering payloads in Astarte Channels. The format of the payload
+can be found in the [Default action](#default-action) section.
+
+Astarte also provides a powerful templating mechanism for plain-text payloads based on top of
+[Mustache](https://mustache.github.io/). This is especially useful for integrating with third-party
+actors which require custom plain-text payloads. Keep in mind that Mustache templates are only able
+to produce `text/plain` payloads, not JSON.
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: not *valid* JSON.

File: doc/pages/architecture/060-triggers.md [hunk 14]
```diff
@@ -1,23 +1,376 @@
...
+The payload of the request is JSON document with this format:
+```json
+{
+  "timestamp": "<timestamp>",
+  "device_id": "<device_id>",
+  "event": <event>
+}
+```
+
+`timestamp` is an UTC ISO8601 timestamp (e.g. `"2019-10-16T08:56:08.534377Z"`) representing when the
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: link to ISO8601 spec?

File: doc/pages/architecture/060-triggers.md [hunk 15]
```diff
@@ -1,23 +1,376 @@
...
+  "type": "path_removed",
+  "interface": "<interface>",
+  "path": "<path>"
+}
+```
+
+`interface` is the interface on which data was received.
+
+`path` is the path that has been removed.
+
```
Comment (bettio) [inline doc/pages/architecture/060-triggers.md]: what about binary blobs? how are they encoded?

Comment (bettio) [reply doc/pages/architecture/060-triggers.md]: are we using any special encoding for big integers?

File: doc/pages/user/060-using_triggers.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+Assuming the above JSON is contained in `my_connection_trigger.json`, you can install the trigger
+using astartectl:
+
+```
+astartectl realm-management triggers install my_connection_trigger.json \
+  -k <realm-key> -r <realm_name> -u <astarte-api-url>
+```
+
```
Comment (bettio) [inline doc/pages/user/060-using_triggers.md]: You can also install the trigger using realm management REST API.

**Pull request merged**